File: src/main/java/com/indiepost/service/CampaignServiceImpl.java
package com.indiepost.service;

import com.indiepost.dto.analytics.*;
import com.indiepost.enums.Types;
import com.indiepost.model.Banner;
import com.indiepost.model.analytics.Campaign;
import com.indiepost.model.analytics.Link;
import com.indiepost.repository.CampaignRepository;
import com.indiepost.repository.ClickRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import javax.inject.Inject;
import javax.transaction.Transactional;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static com.indiepost.utils.DateUtil.*;

/**
 * Created by jake on 8/10/17.
 */
@Service
@Transactional
public class CampaignServiceImpl implements CampaignService {

    private final CampaignRepository campaignRepository;

    private final ClickRepository clickRepository;

    @Inject
    public CampaignServiceImpl(CampaignRepository campaignRepository,
                               ClickRepository clickRepository) {
        this.campaignRepository = campaignRepository;
        this.clickRepository = clickRepository;
    }

    @Override
    public Long save(CampaignDto campaignDto) {
        Campaign campaign = new Campaign();
        campaign.setName(campaignDto.getName());
        campaign.setClientName(campaignDto.getClientName());
        campaign.setGoal(campaignDto.getGoal());
        campaign.setStartAt(instantToLocalDateTime(campaignDto.getStartAt()));
        campaign.setEndAt(instantToLocalDateTime(campaignDto.getEndAt()));
        campaign.setCreatedAt(LocalDateTime.now());
        List<LinkDto> dtoList = campaignDto.getLinks();
        if (dtoList != null && dtoList.size() > 0) {
            List<Link> newLinks = dtoList.stream()
                    .filter(dto -> dto.getId() == null)
                    .map(dto -> linkDtoToNewLink(dto, campaign))
                    .collect(Collectors.toList());
            campaign.setLinks(newLinks);
        }
        campaignRepository.save(campaign);
        return campaign.getId();
    }

    @Override
    public void update(CampaignDto campaignDto) {
        Optional<Campaign> optionalCampaign = campaignRepository.findOne(campaignDto.getId());
        if (!optionalCampaign.isPresent()) {
            //TODO
            return;
        }
        Campaign campaign = optionalCampaign.get();
        campaign.setName(campaignDto.getName());
        campaign.setClientName(campaignDto.getClientName());
        campaign.setGoal(campaignDto.getGoal());
        campaign.setStartAt(instantToLocalDateTime(campaignDto.getStartAt()));
        campaign.setEndAt(instantToLocalDateTime(campaignDto.getEndAt()));
        campaign.setCreatedAt(LocalDateTime.now());
        List<LinkDto> dtoList = campaignDto.getLinks();
        if (dtoList == null || dtoList.size() == 0) {
            return;
        }
        List<LinkDto> updatedLinks = dtoList.stream()
                .filter(dto -> dto.isUpdated())
                .collect(Collectors.toList());
        for (LinkDto updatedLink : updatedLinks) {
            for (Link link : campaign.getLinks()) {
                if (link.getId().equals(updatedLink.getId())) {
                    link.setName(updatedLink.getName());
                    link.setUrl(updatedLink.getUrl());
                }
            }

        }
        List<Link> newLinks = dtoList.stream()
                .filter(dto -> dto.getId() == null)
                .map(dto -> linkDtoToNewLink(dto, campaign))
                .collect(Collectors.toList());
        if (newLinks.size() > 0) {
            campaign.getLinks().addAll(newLinks);
        }
    }

    @Override
    public void deleteById(Long id) {
        campaignRepository.deleteById(id);
    }

    @Override
    public CampaignDto findById(Long id) {
        Optional<Campaign> optional = campaignRepository.findOne(id);
        if (optional.isPresent()) {
            Campaign campaign = optional.get();
            return campaignToDto(campaign, true);
        } else {
            return null;
        }
    }

    @Override
    public Page<CampaignDto> find(Pageable pageable) {
        Page<Campaign> result = campaignRepository.find(pageable);
        List<CampaignDto> dtoList = result.getContent().stream()
                .map(campaign -> campaignToDto(campaign))
                .collect(Collectors.toList());
        return new PageImpl<>(dtoList, result.getPageable(), result.getTotalElements());
    }

    @Override
    public Campaign dtoToCampaign(CampaignDto campaignDto) {
        Campaign campaign = new Campaign();
        campaign.setName(campaignDto.getName());
        campaign.setClientName(campaignDto.getClientName());
        campaign.setGoal(campaignDto.getGoal());
        campaign.setStartAt(instantToLocalDateTime(campaignDto.getStartAt()));
        campaign.setEndAt(instantToLocalDateTime(campaignDto.getEndAt()));
        campaign.setCreatedAt(LocalDateTime.now());
        return campaign;
    }

    @Override
    public List<BannerDto> findBannersOnCampaignPeriod() {
        List<Banner> bannerList = campaignRepository.findBannerOnCampaignPeriodByPriority();
        if (bannerList == null) {
            return new ArrayList<>();
        }
        String linkUriPath = "https://www.indiepost.co.kr/link/";
        return bannerList.stream().map(banner -> {
            BannerDto dto = new BannerDto();
            dto.setId(banner.getId());
            dto.setTitle(banner.getTitle());
            dto.setSubtitle(banner.getSubtitle());
            dto.setBannerType(banner.getBannerType().toString());
            dto.setBgColor(banner.getBgColor());
            dto.setInternalUrl(banner.getInternalUrl());
            dto.setImageUrl(banner.getImageUrl());
            dto.setInternalUrl(banner.getInternalUrl());
            dto.setCover(banner.isCover());
            dto.setPriority(banner.getPriority());
            Link link = banner.getLink();
            if (link != null) {
                dto.setLinkTo(linkUriPath + link.getUid());
            }
            return dto;
        }).collect(Collectors.toList());
    }

    @Override
    public CampaignDto campaignToDto(Campaign campaign) {
        return this.campaignToDto(campaign, false);
    }

    @Override
    public CampaignDto campaignToDto(Campaign campaign, boolean withLinks) {
        CampaignDto dto = new CampaignDto();
        dto.setId(campaign.getId());
        dto.setName(campaign.getName());
        dto.setClientName(campaign.getClientName());
        dto.setCreatedAt(campaign.getCreatedAt());
        dto.setStartAt(localDateTimeToInstant(campaign.getStartAt()));
        dto.setEndAt(localDateTimeToInstant(campaign.getEndAt()));
        dto.setGoal(campaign.getGoal());

        Long campaignId = campaign.getId();
        Long validClick = clickRepository.countValidClicksByCampaignId(campaignId);
        dto.setValidClicks(validClick);
        if (withLinks) {
            Long allClicks = clickRepository.countAllClicksByCampaignId(campaignId);
            dto.setAllClicks(allClicks);
            List<LinkDto> links = campaignRepository.findCampaignLinksOrderByClicks(campaignId);
            dto.setLinks(links);
        }
        return dto;
    }

    @Override
    public CampaignReport getReport(Long id) {
        Optional<Campaign> optional = campaignRepository.findOne(id);
        if (!optional.isPresent()) {
            return new CampaignReport();
        }
        CampaignDto dto = campaignToDto(optional.get(), true);
        List<ShareStat> byLinks = dto.getLinks()
                .stream()
                .map(link -> new ShareStat(link.getName(), link.getValidClicks()))
                .collect(Collectors.toList());
        dto.setLinks(null);

        LocalDateTime start = instantToLocalDateTime(dto.getStartAt());
        LocalDateTime end = instantToLocalDateTime(dto.getEndAt());
        List<TimeDomainStat> trend = campaignRepository.getUniqueVisitorTrendHourly(id);
        trend = normalizeHoursOFTimeDomainStats(trend, start.toLocalDate(), end.toLocalDate());
        List<ShareStat> byOs = campaignRepository.getUniqueVisitorGroupByOs(id);
        List<ShareStat> byBrowsers = campaignRepository.getUniqueVisitorGroupByBrowser(id);

        LocalDateTime since = start.minusDays(60);
        List<ShareStat> topPrevious = campaignRepository.getTopCampaigns(since, end, dto.getClientName(), 10);
        List<RawDataReportRow> rawData = campaignRepository.getRawDataReport(id);

        CampaignReport report = new CampaignReport();
        report.setByOS(byOs);
        report.setByLinks(byLinks);
        report.setByBrowsers(byBrowsers);
        report.setTopPrevious(topPrevious);
        report.setTrend(trend);
        report.setCampaign(dto);
        report.setRawData(rawData);
        return report;
    }

    private Link linkDtoToNewLink(LinkDto dto, Campaign campaign) {
        Link link = new Link();
        link.setCampaign(campaign);
        link.setCreatedAt(LocalDateTime.now());
        link.setName(dto.getName());
        link.setUid(dto.getUid());
        link.setUrl(dto.getUrl());
        if (dto.getBanner() != null) {
            BannerDto bannerDto = dto.getBanner();
            Banner banner = new Banner();
            banner.setTitle(bannerDto.getTitle());
            banner.setSubtitle(bannerDto.getSubtitle());
            banner.setBannerType(Types.BannerType.valueOf(bannerDto.getBannerType()));
            banner.setBgColor(bannerDto.getBgColor());
            banner.setInternalUrl(bannerDto.getInternalUrl());
            banner.setImageUrl(bannerDto.getImageUrl());
            banner.setPriority(bannerDto.getPriority());
            link.setBanner(banner);
        }

        return link;
    }
}

File: src/main/java/com/indiepost/service/CampaignServiceImpl.kt
package com.indiepost.service

import com.indiepost.dto.analytics.*
import com.indiepost.enums.Types
import com.indiepost.model.Banner
import com.indiepost.model.analytics.Campaign
import com.indiepost.model.analytics.Link
import com.indiepost.repository.CampaignRepository
import com.indiepost.repository.ClickRepository
import com.indiepost.utils.DateUtil.*
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.util.stream.Collectors
import javax.inject.Inject
import javax.persistence.EntityNotFoundException
import javax.transaction.Transactional

/**
 * Created by jake on 8/10/17.
 */
@Service
@Transactional
class CampaignServiceImpl @Inject constructor(
        private val campaignRepository: CampaignRepository,
        private val clickRepository: ClickRepository) : CampaignService {

    override fun save(campaignDto: CampaignDto): Long? {
        val campaign = Campaign()
        campaign.name = campaignDto.name
        campaign.clientName = campaignDto.clientName
        campaign.goal = campaignDto.goal
        campaign.startAt = instantToLocalDateTime(campaignDto.startAt!!)
        campaign.endAt = instantToLocalDateTime(campaignDto.endAt!!)
        campaign.createdAt = LocalDateTime.now()
        val dtoList = campaignDto.links
        if (dtoList.isNotEmpty()) {
            val newLinks = dtoList.stream()
                    .filter { (id) -> id == null }
                    .map { dto -> linkDtoToNewLink(dto, campaign) }
                    .collect(Collectors.toList())
            campaign.links = newLinks
        }
        campaignRepository.save(campaign)
        return campaign.id
    }

    override fun update(campaignDto: CampaignDto) {
        val campaign = campaignRepository.findOne(campaignDto.id!!)
                ?: throw EntityNotFoundException("No campaign with this id: " + campaignDto.id!!)
        campaign.name = campaignDto.name
        campaign.clientName = campaignDto.clientName
        campaign.goal = campaignDto.goal
        campaign.startAt = instantToLocalDateTime(campaignDto.startAt!!)
        campaign.endAt = instantToLocalDateTime(campaignDto.endAt!!)
        campaign.createdAt = LocalDateTime.now()
        val dtoList = campaignDto.links
        if (dtoList.isEmpty()) {
            return
        }
        val updatedLinks = dtoList.stream()
                .filter { (_, _, _, _, _, _, _, _, _, _, isUpdated) -> isUpdated }
                .collect(Collectors.toList())
        for ((id, name, url) in updatedLinks) {
            for (link in campaign.links) {
                if (link.id == id) {
                    link.name = name
                    link.url = url
                }
            }

        }
        val newLinks = dtoList.stream()
                .filter { (id) -> id == null }
                .map { dto -> linkDtoToNewLink(dto, campaign) }
                .collect(Collectors.toList())
        if (newLinks.size > 0) {
            campaign.links.addAll(newLinks)
        }
    }

    override fun deleteById(id: Long?) {
        campaignRepository.deleteById(id!!)
    }

    override fun findById(id: Long?): CampaignDto? {
        val campaign = campaignRepository.findOne(id!!)
        return if (campaign != null) {
            campaignToDto(campaign, true)
        } else {
            null
        }
    }

    override fun find(pageable: Pageable): Page<CampaignDto> {
        val result = campaignRepository.find(pageable)
        val dtoList = result.content.stream()
                .map { campaign -> campaignToDto(campaign) }
                .collect(Collectors.toList())
        return PageImpl(dtoList, result.pageable, result.totalElements)
    }

    override fun dtoToCampaign(campaignDto: CampaignDto): Campaign {
        val campaign = Campaign()
        campaign.name = campaignDto.name
        campaign.clientName = campaignDto.clientName
        campaign.goal = campaignDto.goal
        campaign.startAt = instantToLocalDateTime(campaignDto.startAt!!)
        campaign.endAt = instantToLocalDateTime(campaignDto.endAt!!)
        campaign.createdAt = LocalDateTime.now()
        return campaign
    }

    override fun findBannersOnCampaignPeriod(): List<BannerDto> {
        val bannerList = campaignRepository.findBannerOnCampaignPeriodByPriority()
        val linkUriPath = "https://www.indiepost.co.kr/link/"
        return bannerList.stream().map { banner ->
            val dto = BannerDto()
            dto.id = banner.id
            dto.title = banner.title
            dto.subtitle = banner.subtitle
            dto.bannerType = banner.bannerType!!.toString()
            banner.bgColor?.let {
                dto.bgColor = it
            }
            dto.internalUrl = banner.internalUrl
            dto.imageUrl = banner.imageUrl
            dto.internalUrl = banner.internalUrl
            dto.isCover = banner.isCover
            dto.priority = banner.priority
            dto.target = banner.target.toString()
            val link = banner.link
            if (link != null) {
                dto.linkTo = linkUriPath + link.uid!!
            }
            dto
        }.collect(Collectors.toList())
    }

    override fun campaignToDto(campaign: Campaign): CampaignDto {
        return this.campaignToDto(campaign, false)
    }

    override fun campaignToDto(campaign: Campaign, withLinks: Boolean): CampaignDto {
        val dto = CampaignDto()
        dto.id = campaign.id
        dto.name = campaign.name
        dto.clientName = campaign.clientName
        dto.createdAt = campaign.createdAt
        dto.startAt = localDateTimeToInstant(campaign.startAt!!)
        dto.endAt = localDateTimeToInstant(campaign.endAt!!)
        dto.goal = campaign.goal

        val campaignId = campaign.id
        val validClick = clickRepository.countValidClicksByCampaignId(campaignId!!)
        dto.validClicks = validClick
        if (withLinks) {
            val allClicks = clickRepository.countAllClicksByCampaignId(campaignId)
            dto.allClicks = allClicks
            val links = campaignRepository.findCampaignLinksOrderByClicks(campaignId)
            dto.links = links
        }
        return dto
    }

    override fun getReport(id: Long?): CampaignReport {
        val campaign = campaignRepository.findOne(id!!) ?: return CampaignReport()
        val dto = campaignToDto(campaign, true)
        val byLinks = dto.links
                .stream()
                .map { (_, name, _, _, _, _, _, _, validClicks) -> ShareStat(name!!, validClicks!!) }
                .collect(Collectors.toList())

        val start = instantToLocalDateTime(dto.startAt!!)
        val end = instantToLocalDateTime(dto.endAt!!)
        var trend = campaignRepository.getUniqueVisitorTrendHourly(id)
        trend = normalizeHoursOFTimeDomainStats(trend, start.toLocalDate(), end.toLocalDate())
        val byOs = campaignRepository.getUniqueVisitorGroupByOs(id)
        val byBrowsers = campaignRepository.getUniqueVisitorGroupByBrowser(id)

        val since = start.minusDays(60)
        val topPrevious = campaignRepository.getTopCampaigns(since, end, dto.clientName!!, 10)
        val rawData = campaignRepository.getRawDataReport(id)

        val report = CampaignReport()
        report.byOS = byOs
        report.byLinks = byLinks
        report.byBrowsers = byBrowsers
        report.topPrevious = topPrevious
        report.trend = trend
        report.campaign = dto
        report.rawData = rawData
        return report
    }

    private fun linkDtoToNewLink(dto: LinkDto, campaign: Campaign): Link {
        val link = Link()
        link.campaign = campaign
        link.createdAt = LocalDateTime.now()
        link.name = dto.name
        link.uid = dto.uid
        link.url = dto.url
        if (dto.banner != null) {
            val bannerDto = dto.banner
            val banner = Banner()
            banner.title = bannerDto!!.title
            banner.subtitle = bannerDto.subtitle
            bannerDto.bannerType?.let {
                banner.bannerType = Types.BannerType.valueOf(it)
            }
            banner.bgColor = bannerDto.bgColor
            banner.internalUrl = bannerDto.internalUrl
            banner.imageUrl = bannerDto.imageUrl
            banner.priority = bannerDto.priority
            link.banner = banner
        }
        return link
    }
}
